feat: Add SSR feature

thanhpd opened the pull request:
### Description
Add SSR feature. Notable changes include:
- Added and updated files following [universal-starter](https://github.com/angular/universal-starter) project
- Updated npm packages (See below for the explanation)
- Changed paths in TypeScript config to relative paths. (See below for explanation)
- Added `Window` service for SSR build
- Added Mock service for `AlgoliaService`, `NlpService` for SSR build
- Added Firebase deployment for SSR using Firebase hosting and functions
- Cleaned up unused imports
- Updated build and deploy sections in documentation

**Updated npm packages**
- Using HEAD version for Angular packages caused errors so I have to change to fix versions.

**Change paths in TypeScript config**
- Using `baseUrl` with `paths` in `tsconfig.json` will produce the error:
```
ERROR in src\app\app.module.ts(18,5): Error during template compile of 'AppModule'
  Could not resolve @app/core/core.module relative to C:/Users/thanh/Source/Repos/angular-search-experience2/src/app/app.module.ts.
```

### Preview
- Client-side version: `https://ecstatic-thompson-ca6f36.netlify.com/`
- Firebase with SSR: `https://thanhpd-aes.firebaseapp.com/`

### Note
This PR is related to #9

Commit:
initial commit

Commit:
fix: clean the cli (v6.0.0-rc.4) mis-generated blueprint

Commit:
chore: switch to yarn

Commit:
build: ignore compodoc output directory

Commit:
chore: prepare a custom material design theme

Commit:
chore: use typescript path aliases for a better flexibilty

Commit:
chore: add a dataset of mocks

Commit:
feat: add basic modules and components of the app shell

- app.module: angular's root module
- app-routing.module: angular's root module's routes definition
- app.component: root component
- core/material.module: core material holding Material dependencies
- search-ui/**: the search feature module (lazy loaded)

Commit:
build: add hammer dependency (for material)

Commit:
chore: move the mocks dataset to the assets folder

Commit:
chore: add material icons font

Commit:
chore: import missing hammer dependency

Commit:
revert: custom theme

Commit:
chore: add more material modules

Commit:
chore: add support for local search (still prototyping)

Commit:
chore: add background image + header gradient

Commit:
chore: move the header div to app.component

Commit:
chore: improve the UI of the search parent component

- wrap the search input and result inside a mat-card
- hook the search service (with local mock data)

Commit:
chore: implement the search input UI and logic

Commit:
chore: implement the UI and logic of the search result component

Commit:
chore: import the model driven forms module

Commit:
build: add firebase config for hosting

Commit:
style: apply prettier formatting to tsconfig

Commit:
fix: Fix AOT compilation errors (language service seems to be broken).

Commit:
docs: add credits information (for images)

Commit:
chore: add Algolia credits

Commit:
fix: set the correct firebase public folder

Commit:
fix: algolia's logo position

Commit:
chore: add easter ascii egg

Commit:
refactor: being more reactive

Commit:
chore: move app shell styles and content to index.html for faster FMP

Commit:
chore: limit mocks to 100 first

Commit:
refactor: redesign components communication for better maintainability

Commit:
feat(server): scaffold cloud functions

Commit:
chore(server): freeze Node version required by firebase (v6.11.5)

Commit:
chore(server): add express dependencies

Commit:
chore(server): add initial function implementation

Commit:
chore(server): add dotenv config + bump dependencies

Commit:
chore(server): implement POST and DELETE objects

Commit:
chore(server): switch to yarn

Commit:
chore: freeze Node version for the UI

Commit:
build: add algoliasearch npm dependencies

Commit:
chore: configure algolia application and search API keys

Commit:
chore: update page title

Commit:
feat: add Algolia NgModule

Commit:
chore: use the new Algolia NgModule

Commit:
feat: add rating directive

Commit:
chore: make the background image contains

Commit:
feat: show the apps price

Commit:
chore: format price. Set to "free" is 0.00 USD

Commit:
chore: add link to algolia

Commit:
chore: center the title inside the main header...

and remove the padding-bottom in order to avoid showing the scrolling bar

Commit:
chore: leave some room at the bottom of the page

Commit:
refactor: reorder imports

Commit:
feat: add a custom infinite scrolling directive

Commit:
chore: implement the infinite scrolling behavior

Commit:
refactor: remove logs

Commit:
chore: leave some space after the title in the header

Commit:
chore: return unsubsciption for algolia events

Commit:
feat: add UI disjunctive facets on "category"

Commit:
fix: use the whole width for the input field

Commit:
fix: reset the facet state when results are found

Commit:
chore: add page loading indicator

Commit:
chore: rename algolia's search index

Commit:
style: apply linter

Commit:
chore: remove currency pipe now that the price is a number

Commit:
chore: auto focus the search field on app load

Commit:
chore: rework the indeces and replicas (necessary for sorting strategies)

Commit:
chore: trigger infinite scrolling before reaching the bottom of the viewport

Commit:
feat: add sorting support:

- most relevant (index: applications)
- lowest price (index: applications_by_price_asc)
- most popular (index: applications_by_rating_desc)

Commit:
chore: remove logs

Commit:
refactor: remove all spec files (sorry)

Commit:
feat: add support for attributes highlights

Commit:
chore: add angular logo

Commit:
feat: add supprot for deeplinking

Commit:
docs: add COC

Commit:
docs: add CONTRIBUTION guide

Commit:
docs: add LICENSE

Commit:
chore: move to css BEM naming

Commit:
chore: add gtag tracking code

Commit:
build: add firebase-tools as a local dependency

Commit:
doc: generate API documentation

Commit:
build: add "doc" script to generate API documentation

Commit:
doc: add initial HOWTO documentation

Commit:
docs: use 'docs' forlder for github pages

Commit:
docs: add screenshot

Commit:
docs: reference the screenshot in README.md

Commit:
docs: fix the screenshot url

Commit:
docs: center the screenshot

Commit:
docs: update HOWTO guide

Commit:
docs: bigger screenshot

Commit:
docs: update API docs

Commit:
docs: update readme

Commit:
docs: add a curl POST command example

Commit:
docs: add algolia logo

Commit:
docs: add a DELETE cURL example

Commit:
docs: add an Architecture section

Commit:
docs: add deeplink and lazy loading gifs

Commit:
fix: rename facets to categories

Commit:
Merge branch 'master' of github.com:manekinekko/angular-search-experience

Commit:
ci: add circleci config

Commit:
chore: add cloud function custom redirects

Commit:
chore: update firebase redirects

Commit:
fix: circleci test command

Commit:
chore: disable tests on circleci due to missing chrome

Commit:
feat(backend): secure the api with helmet

Commit:
fix: failind test command on circleci

Commit:
docs: use the new domaine

Commit:
docs: add circleci badge

Commit:
docs: add token API to badge

Commit:
docs: add launch button

Commit:
docs: update title

Commit:
docs: add composition diagram

Commit:
docs: add component architecture

Commit:
docs: add gifs for search results and search box

Commit:
feat: eaaaaaaaaaaaaster egg 😇

Commit:
Merge branch 'master' of github.com:manekinekko/angular-search-experience

Commit:
fix: remove applications count

Commit:
feat: add pwa'ness

Commit:
fix: observable imports

Commit:
fix: pwa icons path

Commit:
refactor: easter egg

Commit:
chore(backend): easter egg

Commit:
docs: update api docs

Commit:
docs: fix images width to 800

Commit:
docs: update api documentation

Commit:
fix: remove plurial

Commit:
docs: update the API documentation

Commit:
chore: remove 'angular' from the title

Commit:
chore: add sketch src files

Commit:
chore: add a custom logo

Commit:
fix: no-result background was missing

Commit:
chore: bump angular to 6.0.0-rc.5

Commit:
chore: add credits in footer

Commit:
fix: easter

Commit:
docs: add easter egg instructions

Commit:
docs: add disclaimer

Commit:
chore: add a firebase functions deploy script

Commit:
chore: improving the egg

Commit:
fix: infinite scrolling

Commit:
fix: remove unused operators

Commit:
docs: add changelog

Commit:
docs: update api documentation

Commit:
chore(server): add a local mock server for non-firebase users

Commit:
docs: update docs to use the local experess server

Commit:
docs: add missing env variables

Commit:
docs: updated readme

Commit:
docs: update github account

Commit:
Merge pull request #2 from sfeir-open-source/manekinekko-patch-1

docs: update github account

Commit:
doc: add a link to Angular InstalSearch

Commit:
Merge pull request #4 from sfeir-open-source/manekinekko-patch-1

doc: add a link to Angular InstalSearch

Commit:
chore: fix the broken algolia logo

Commit:
chore: sort categories to avoid undesired unordering

Commit:
chore: update to angular 6.1.1 and material 6.4.2 + move to npm

Commit:
docs: update docs link

SiegfriedEhret committed:
test: fix e2e test (#7)

Commit:
docs: add a CoC

manekinekko committed:
docs: Update issue templates

manekinekko committed:
docs: create a PR template

thanhpd committed:
feat: Add SSR feature
Close #9

manekinekko commented:
This PR should be merged after https://github.com/sfeir-open-source/angular-search-experience/pull/14

tinesoft committed:
chore: refactor master builds (#14)

* refactor: update angular versions to master builds

* build: fix compilation error due to `new window[webkitSpeechRecognition` usage

* build: fix compilation error due to conflicting `api-ai-javascript` typings (with `lib.dom.d.ts`)

thanhpd commented:
@manekinekko 
I've updated the PR and descriptions with previews, please take a look.

wolfgangGoedel committed:
refactor(types): use union type for range (#20)

wolfgangGoedel committed:
fix(devServer): log correct port when set in env (#16)

alexbrbr committed:
chore(install): use npm instead of yarn #12 (#18)

thanhpd commented:
Rebased PR
@manekinekko This PR is completed, the build is failing because I changed the `outputPath` to differentiate between client and server bundles location. Please take a look.
